File: Python3/1 - 200/124 Binary Tree Maximum Path Sum.py
# 88ms, 77.34%; 19.5MB, 100.00%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.max = float('-inf')
        self.helper(root)
        return self.max

    def helper(self, root: TreeNode) -> int:
        if not root: return 0
        left, right = self.helper(root.left), self.helper(root.right)
        _sum = root.val + max(0, left) + max(0, right)
        self.max = max(self.max, _sum)
        return root.val + max(0, left, right)

File: 1 - 200/124 Binary Tree Maximum Path Sum.cpp
// 24ms, 96.53%; 24.6MB, 72.73%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
private:
    int _max = INT_MIN;

public:
    int maxPathSum(TreeNode *root)
    {
        helper(root);
        return _max;
    }

    int helper(TreeNode *root)
    { // update `_max`, `root->val` itself is always included
        if (!root) return 0;
        int left = helper(root->left), right = helper(root->right);
        int _sum = root->val + max(0, left) + max(0, right);
        _max = max(_max, _sum);
        return root->val + max({0, left, right});
    }
};

